docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+      - [Story 3](#story-3)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+* **Node Group** This term has special meaning within the cluster autoscaler, it refers to collections
+  of nodes, and related physical hardware, that are organized within the autoscaler for scaling operations.
+  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled
+  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond
+  directly to MachineSets and MachineDeployments that are marked for autoscaling.
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters for MachineSets and MachineDeployments. One feature
+that is missing from this integration is the ability to scale down to, and up from, a MachineSet
+or MachineDeployment with zero replicas.
+
+This proposal describes opt-in mechanisms whereby Cluster API users and infrastructure providers can define
+the specific resource requirements for each Infrastructure Machine Template they create. In situations
+where there are zero nodes in the node group, and thus the autoscaler does not have information
+about the nodes, the resource requests are utilized to predict the number of nodes needed. This
+information is only used by the autoscaler when it is scaling from zero.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated infrastructure providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and infrastructure providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
```

Comment:
nit: Provide capability for Cluster API MachineSets and MachineDeployments to be auto scaled from and to zero replicas by the cluster autoscaler.

Scaling from/to zero expressing intent, is possible today.